### DISCUSSION TITLE
Easiest way to cache a resource's results?

### DISCUSSION THREAD

**Discussion by blumbra63** [leptos-rs/leptos#1162] [Q&A] [answered]
Hey all. I've been playing around with porting a very simple WIP web app from Yew to Leptos. I'm most of the way there, but I've found that something which was simple in Yew, is not so simple in Leptos. Specifically, I have an API that returns one month's worth of data, and the UI is such that users can select a month to view. I'm looking to cache the results so if the user chooses month A, month B, then month A again, the same data is reused.

In Yew today, I have a `HashMap` state variable, and an effect that triggers based on a selected month state variable. The effect checks if the value corresponding to the month is populated in the `HashMap`, and if it is, it returns it; otherwise, it `await`s on the API request, places that value in the `HashMap`, and returns it.

When I tried to port that same code over to Leptos, making the `HashMap` a signal and that effect a `Resource`, the framework was unhappy with me for modifying the signal's value inside the resource, because it is "not a reactive context". It's possible I'm doing something wrong, but I also have the sense that there is a better/easier way to do this in Leptos. Would anyone be able to give me a hand?

**Comment by blumbra63** [comment]
I've got something which seems to work, but feels fragile. It has three parts:
- A memo storing a tuple of the set of the keys requested, and the last key requested. Note the last key is only set if it is new to the set.
- A resource that depends on the last key requested, and fetches the value.
- A memo that is a HashMap of keys to values, that depends on the resource.

The duplication between the first and last memo causes me unease, but this seems to work, despite being somewhat convoluted. I am all ears if there are any better ideas, or if I am way off base.

**Comment by gbj** [comment, marked as answer]
Why not just have a `HashMap` where the key is the month and the value is a resource of the data for that month? It doesn't even seem like it needs to be reactive, to me; just an `Rc<RefCell<HashMap<_, _>>>` or, for convenience, use `store_value` and have a `StoredValue<HashMap<_, _>>`? Whenever you switch months, you then use the map's entry API to either get the corresponding resource or create a new resource.

**Comment by blumbra63** [comment]
Thanks for the suggestion. I think I got so pigeon-holed into thinking in terms of reactive elements, that I forgot I could code without them... Code seems much less brittle now. :)
